### DISCUSSION TITLE
JSBSim Increase Output Directive Precision

### DISCUSSION THREAD

**Discussion by sr71684** [JSBSim-Team/jsbsim#528] [Q&A]
**Describe the issue**

How do I increase the precision of JSBSim output directive properties? 

**What is the current behavior?**
Precision of `position/lat-geod-deg` and `position/long-gc-deg` is very low, at around 4-5 significant figures.  

**What is the expected behavior?**

I want to increase precision to at least 16 significant figures.

**Comment by seanmcleod** [comment]
Can you post a copy of how you're specifying your outputs?

If you're outputting to a text file then it looks like by default the precision is set to 10.

https://github.com/JSBSim-Team/jsbsim/blob/dd97e99032e6cce6ae3d0d91730f3c10f8ce94af/src/input_output/FGOutputTextFile.cpp#L260

However for certain categories of properties the precision is increased, e.g. 

https://github.com/JSBSim-Team/jsbsim/blob/dd97e99032e6cce6ae3d0d91730f3c10f8ce94af/src/input_output/FGOutputTextFile.cpp#L341-L364

If you specify individual properties as opposed to a category like `ssPropogate` then it looks like the precision is set to 18.

https://github.com/JSBSim-Team/jsbsim/blob/dd97e99032e6cce6ae3d0d91730f3c10f8ce94af/src/input_output/FGOutputTextFile.cpp#L382-L385

**Comment by bcoconni** [comment]
This topic has been moved to JSBSim Discussions as per the issue template statement: `Issues that are neither bug reports or feature requests will be transferred to [Discussions](https://github.com/JSBSim-Team/jsbsim/discussions) by the team and the issue will be locked and closed.`.

Support for this topic will therefore continue within this discussion. Thanks for your understanding.

**Reply by CarloBlanch1967** [reply]
Hello, I'm experiencing the same issue, but on socket via udp.
Precision of  position/lat-geod-deg and position/long-gc-deg is very low, only 5 decimal digits.

How can I increase them?


Thanks,

**Comment by seanmcleod** [comment]
@CarloBlanch1967 have you looked at the source code to confirm whether the precision is hard-coded for the UDP output option that you're using?

I took a look at the source for the text output option in the comment above.

**Reply by CarloBlanch1967** [reply]
Yes, but I didn't see anything similar in FGOutputSocket.cpp

Thanks a lot.

**Comment by seanmcleod** [comment]
So for doubles the field width is hard-coded to 12 and the precision is hard-coded to 7. 

https://github.com/JSBSim-Team/jsbsim/blob/60041d7ee2e30ee961bd9f491e8a92be24e7de60/src/input_output/FGOutputSocket.cpp#L359-L360

https://github.com/JSBSim-Team/jsbsim/blob/60041d7ee2e30ee961bd9f491e8a92be24e7de60/src/input_output/FGfdmSocket.cpp#L338-L350

**Reply by CarloBlanch1967** [reply]
Thanks a lot!

Worderful work, and quick assistance!

Regards,

**Comment by seanmcleod** [comment]
Just to be clear, the only work I've done so far is to double check the existing code for the socket option to confirm if the precision is hard-coded and if so what it's currently set to.

One option to allow users to set the precision used by the socket output is to add support for a new element, e.g. `precision` along these lines.

```xml
<output name="localhost" type="SOCKET" port="5138" protocol="UDP" rate="10" precision="8">
   
  <property> velocities/vc-kts </property>
  <property> position/h-sl-ft </property>
  <property> attitude/theta-deg </property>

</output>
```

**Reply by CarloBlanch1967** [reply]
OMG,  I am completely stupid or this parameter is not documented?
Maybe the first,  but I'm very happy , this solved my issue.

Thanks again.

**Comment by seanmcleod** [comment]
No, you've misunderstood what I was saying. I wasn't saying there is such an option currently.

I was saying that one option we could implement/add "..is to add support for a new element..." is to write some code that would support this new element.

So to be clear, nothing has been solved.

**Reply by CarloBlanch1967** [reply]
Ok, sorry.
Yes, It would be a good solution.
Thanks again!

**Comment by seanmcleod** [comment]
@CarloBlanch1967 I've just submitted a pull request - https://github.com/JSBSim-Team/jsbsim/pull/579 to allow a user to optionally specify the output precision on output sockets.

**Reply by dpculp** [reply]
I'd like to thank Sean and Bertrand for this.  I've been having a loss-of-precision problem with QtJSBSim and couldn't find the source.  I had been assuming the loss occurred in the JavaScript, but I was wrong!

With 7 significant digits I was getting latitude and longitude jitters of about 3 to 30 feet respectively.  It wasn't noticeable  until I used an external viewer like Cesium or FlightGear and could see the jitter.

**Reply by bcoconni** [reply]
@dpculp, all the credits should go to @seanmcleod as I did not do anything valuable.
